teps/0076-array-result-types.md

```diff
@@ -0,0 +1,321 @@
+
+For example:
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: Pipeline
+...
+spec:
+  params:
+    - environments:
+        - 'staging'
+        - 'qa'
+        - 'prod'
+...
+  tasks:
+    - name: deploy
+      params:
+        - name: environment
+          value: '$(params.environments[0])'
+```
+
+The resulting taskRun would contain:
+
+```yaml
+  params:
+    - name: environment
+      value: 'staging'
+```
+
+This proposal does not include adding support for any additional syntax, e.g. slicing (though it could be added in the
+future!)
+
+### Using array results with variable replacement
+
+The same syntax [above](#indexing-into-arrays-with-variable-replacement) could be used to access a specific value in
+the results of a previous Task.
+
+For example:
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: Pipeline
+...
+spec:
+tasks:
+  - name: get-environments
+  - name: deploy
+    params:
+      - name: environment
+        value: '$(tasks.get-environments.environments[0])'
+```
+
+## Test Plan
+
+In addition to unit tests the implementation would include:
+
+* At least one reconciler level integration test which includes an array result
+* A tested example of a Task that produces an array result with:
+  * Another Task consuming that array
+  * Another Task consuming one specific value from that array
+
+## Design Evaluation
+
+* [Reusability](https://github.com/tektoncd/community/blob/main/design-principles.md#reusability):
+  * Pro: This will improve the reusability of Tekton components by enabling the scenario of a Task providing an array result
+    and then a Pipeline being able to loop over those value for subsequent Tasks (i.e. reusing Tasks which are designed
+    for single inputs, but users want to use 'for each' item in an array)
+* [Simplicity](https://github.com/tektoncd/community/blob/main/design-principles.md#simplicity)
+  * Pro: This proposal reuses the existing array or string concept for params
+  * Pro: This proposal continues [the precedent of using JSONPath syntax in variable replacement](https://github.com/tektoncd/pipeline/issues/1393#issuecomment-561476075)
+  * Con: There are now two different files/paths to which users can write results and they have to understand when
+    to use which. (We could choose to deprecate the original path if necessary.)
+* [Flexibility](https://github.com/tektoncd/community/blob/main/design-principles.md#flexibility)
+  * Con: Although there is a precedent for including JSONPath syntax, this is a step toward including more hard coded
+    expression syntax in the Pipelines API (without the ability to choose other language options)
+* [Conformance](https://github.com/tektoncd/community/blob/main/design-principles.md#conformance)
+  * Supporting array results and indexing syntax would be included in the conformance surface
+
+## Drawbacks
+
+* Increases the size of TaskRuns (by including arrays in termination messages and results)
```

Comment:
Any possible problem if the arrays is huge (a "from a task" attach that would write a 10M array json file for example) ?

Reply:
if [the termination message ended up being > 4096 bytes it would get truncated](https://kubernetes.io/docs/tasks/debug-application-cluster/determine-reason-pod-failure/#customizing-the-termination-message) (and end up being invalid json and the taskrun would fail)  (unless as suggested in https://github.com/tektoncd/pipeline/issues/4012 we adopt an alternative) - anyway ill mention this in the TEP